libc/string_catdup.c

```diff
@@ -0,0 +1,780 @@
+	}
+}
+
+TEST(string_cat, strncat_ascii)
+{
+	int i;
+	char buf[150] = { 0 };
+	char ascii[128] = { 0 };
+
+	for (i = 0; i < 128; i++) {
+		ascii[i] = i;
+	}
+
+	TEST_ASSERT_EQUAL_PTR(buf, strncat(buf, ascii, i + 1));
+	TEST_ASSERT_EQUAL_STRING(ascii, buf);
+	for (i = strlen(ascii); i < 150; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+}
+
+
+TEST(string_cat, strncat_extended_ascii)
+{
+	int i;
+	char buf[150] = { 0 };
+	char ascii[128] = { 0 };
+
+	for (i = 128; i < 255; i++) {
+		ascii[i - 128] = i;
+	}
+
+	TEST_ASSERT_EQUAL_PTR(buf, strncat(buf, ascii, i - 127));
+	TEST_ASSERT_EQUAL_STRING(ascii, buf);
+	for (i = strlen(ascii); i < 150; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+}
+
+
+TEST(string_cat, strncat_non_ascii)
+{
+	int i;
+	char buf[SIZE_BUF * 2] = NON_ASCII_1;
+	char buf1[] = NON_ASCII_2;
+
+	TEST_ASSERT_EQUAL_PTR(buf, strncat(buf, buf1, strlen(NON_ASCII_1 NON_ASCII_2)));
+	TEST_ASSERT_EQUAL_STRING(NON_ASCII_1 NON_ASCII_2, buf);
+
+	for (i = strlen(NON_ASCII_1 NON_ASCII_2); i < SIZE_BUF * 2; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+}
+
+/////////////////////////////////////////////////////////////////////////
+
+TEST_SETUP(string_dup)
+{
+}
+
+
+TEST_TEAR_DOWN(string_dup)
+{
+}
+
+
+TEST(string_dup, strdup_basic)
+{
+	char buf[SIZE_BUF] = WORD_1 WORD_2;
+	char *buf1 = strdup(buf);
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(buf, buf1);
+
+	free(buf1);
+}
+
+
+TEST(string_dup, strdup_empty)
+{
+	char buf[2] = "";
+	char *buf1 = strdup(buf);
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(buf, buf1);
+	TEST_ASSERT_EQUAL_CHAR('\0', buf1[0]);
+
+	free(buf1);
+}
+
+
+TEST(string_dup, strdup_huge_string)
+{
+	int i;
+	char buf[PATH_MAX] = "";
+	char *buf1;
+
+	for (i = 0; i < PATH_MAX - 1; i++) {
+		buf[i] = HUGE_STRING[i];
+	}
+
+	buf1 = strdup(buf);
+
+	TEST_ASSERT_NOT_NULL(buf);
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(buf, buf1);
+
+	free(buf1);
+}
+
+
+TEST(string_dup, strdup_specials_string)
+{
+	char buf[SIZE_BUF * 2] = SPECIALS_STRING;
+	char *buf1 = strdup(buf);
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(buf, buf1);
+
+	free(buf1);
+}
+
+
+TEST(string_cat, strdup_ascii)
+{
+	int i;
+	char *buf;
+	char ascii[128] = { 0 };
+
+	for (i = 0; i < 128; i++) {
+		ascii[i] = i;
+	}
+
+	buf = strdup(ascii);
+
+	TEST_ASSERT_EQUAL_STRING(ascii, buf);
+
+	free(buf);
+}
+
+
+TEST(string_cat, strdup_extended_ascii)
+{
+	int i;
+	char *buf;
+	char ascii[128] = { 0 };
+
+	for (i = 128; i < 255; i++) {
+		ascii[i - 128] = i;
+	}
+
+	buf = strdup(ascii);
+
+	TEST_ASSERT_EQUAL_STRING(ascii, buf);
+
+	free(buf);
+}
+
+
+TEST(string_dup, strdup_non_ascii)
+{
+	char buf[SIZE_BUF * 2] = NON_ASCII_1 NON_ASCII_2;
+	char *buf1 = strdup(buf);
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(buf, buf1);
+
+	free(buf1);
+}
+
+///////////////////////////////////////////////////////////////////////////////////
+
+TEST(string_dup, strndup_basic)
```

Comment:
I would say it's not a basic behavior of this function, but it's a matter of opinion, maybe: 
```suggestion
TEST(string_dup, strndup_part)
```

or sth similar indicating that we duplicate only a part of source string.